File: packages/@bifrostql/types/src/saved-object.ts
/**
 * Saved-object contract — persisted, user-authored objects (queries, forms,
 * reports, dashboards) exposed by the server's `/_saved-objects` endpoint.
 *
 * Mirrors the C# source of truth in `src/BifrostQL.Core/SavedObjects/*.cs`,
 * serialized camelCase with the enum as a camelCase string and nullable fields
 * omitted (`WhenWritingNull`) — so every C# nullable maps to an optional `field?`.
 *
 * Distinct from `SavedViewMetadata` in `./metadata`, which is a read-only grid
 * presentation overlay, not a persisted user object.
 */

/** The kind of user-authored object a {@link SavedObject} holds. */
export type SavedObjectType = 'query' | 'form' | 'report' | 'dashboard';

/**
 * One persisted user-authored object. `definition` is the opaque, type-specific
 * payload the server does not interpret. `version` is an optimistic-concurrency
 * token: a create carries `0` (server persists `1`); an update must carry the
 * version last read, and the server rejects a stale write with HTTP 409.
 */
export interface SavedObject {
  /** Stable client-assigned identifier, unique within its `type`. */
  id: string;
  /** The object kind. */
  type: SavedObjectType;
  /** Human-readable name. */
  name: string;
  /** Optional folder path for organization; omitted = root. */
  folder?: string;
  /** Opaque, type-specific definition payload. */
  definition: unknown;
  /** Optimistic-concurrency version token. */
  version: number;
}

File: src/BifrostQL.Core/SavedObjects/SavedObject.cs
using System.Text.Json;

namespace BifrostQL.Core.SavedObjects;

/// <summary>
/// The kind of user-authored object a <see cref="SavedObject"/> holds. Serialized
/// camelCase over the wire (e.g. <c>"query"</c>); an unrecognized value is rejected
/// by the store rather than silently coerced.
/// </summary>
public enum SavedObjectType
{
    Query,
    Form,
    Report,
    Dashboard,
}

/// <summary>
/// One persisted user-authored object: a query/form/report/dashboard the client
/// designed. <see cref="Definition"/> is the opaque, type-specific payload (a JSON
/// document the server does not interpret). <see cref="Version"/> is an optimistic
/// concurrency token — a write must carry the version it last read, and the store
/// increments it on success and rejects a stale write.
/// </summary>
public sealed record SavedObject
{
    public required string Id { get; init; }
    public required SavedObjectType Type { get; init; }
    public required string Name { get; init; }

    /// <summary>Optional folder path for client-side organization; null = root.</summary>
    public string? Folder { get; init; }

    /// <summary>Opaque, type-specific definition payload. Not interpreted server-side.</summary>
    public required JsonElement Definition { get; init; }

    /// <summary>
    /// Optimistic concurrency token. A create carries 0 (no prior version); the store
    /// persists it as 1. An update carries the version last read; the store writes
    /// version+1 iff it matches the stored version, else raises a conflict.
    /// </summary>
    public int Version { get; init; }
}
